portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -41,18 +45,28 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     PreparedQuery results = datastore.prepare(query);
 
     List<Comment> comments = new ArrayList<>();
+    LanguageServiceClient languageService = LanguageServiceClient.create();
+    float sentimentSum = 0;
+
     for (Entity entity : results.asIterable()) {
       String firstName = (String) entity.getProperty("firstName");
       String lastName = (String) entity.getProperty("lastName");
       String message = (String) entity.getProperty("message");
       long timestamp = (long) entity.getProperty("timestamp");
-
       Comment comment = new Comment(firstName,lastName,message,timestamp);
+
+      Document doc = Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+      Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
```

Comment:
This approach is great (especially since you may have comments already stored), just leaving a comment for you to think about. If you pay for every call to the `languageService` (which I suspect is how the Google ML tools work), then you could probably cache the sentiment scores alongside the comment. Doing that in `doGet` would probably be tricky, but you could store the sentiment when you originally save the comment instead and then the score would just be available when you need it.

But of course! If you start to cache things that introduces its own host of problems to consider. For example, if you gave people the option to edit their comments, then you would need to remember to re-compute the sentiment score every time they save. 

Having said all of this, this code looks good and I don't think you need to change anything :grin: